File: ui/components/ui/LibraryBrowser/model/resource-matchers.ts
type WithReactKey = { reactKey: string };

export type InstanceMatcher = WithReactKey & {
  type: "instance-matcher"
}

export type ExactTenantMatcher = WithReactKey & {
  type: 'exact-tenant-matcher',
  tenant: string;
};

export type AllTenantMatcher = WithReactKey & {
  type: 'all-tenant-matcher'
};

export type TenantMatcher = WithReactKey & {
  type: 'tenant-matcher',
  value: ExactTenantMatcher | AllTenantMatcher
};

export type TenantMatcherType = TenantMatcher['value']['type'];

export type ExactNamespaceMatcher = WithReactKey & {
  type: 'exact-namespace-matcher',
  tenant: TenantMatcher;
  namespace: string;
};

export type AllNamespaceMatcher = WithReactKey & {
  type: 'all-namespace-matcher',
  tenant: TenantMatcher
};

export type NamespaceMatcher = WithReactKey & {
  type: 'namespace-matcher',
  value: ExactNamespaceMatcher | AllNamespaceMatcher
};
export type NamespaceMatcherType = NamespaceMatcher['value']['type'];

export type ExactTopicMatcher = WithReactKey & {
  type: 'exact-topic-matcher',
  namespace: NamespaceMatcher;
  topic: string;
};

export type AllTopicMatcher = WithReactKey & {
  type: 'all-topic-matcher',
  namespace: NamespaceMatcher
};

export type TopicMatcher = WithReactKey & {
  type: 'topic-matcher',
  value: ExactTopicMatcher | AllTopicMatcher
};

export type TopicMatcherType = TopicMatcher['value']['type'];

export type ResourceMatcher = WithReactKey & (InstanceMatcher | TopicMatcher | NamespaceMatcher | TenantMatcher);

File: server/src/main/scala/library/resourceMatchers.scala
package library

case class ExactTenantMatcher(tenant: String):
    def test(target: TenantMatcher): Boolean = target.matcher match
        case t: AllTenantMatcher   => true
        case t: ExactTenantMatcher => t.tenant == tenant

case class AllTenantMatcher():
    def test(target: TenantMatcher): Boolean = true

case class TenantMatcher(matcher: ExactTenantMatcher | AllTenantMatcher):
    def test(target: TenantMatcher): Boolean = matcher match
        case m: ExactTenantMatcher => m.test(target)
        case m: AllTenantMatcher   => m.test(target)

case class AllNamespaceMatcher(tenant: TenantMatcher):
    def test(target: NamespaceMatcher): Boolean = target.matcher match
        case t: AllNamespaceMatcher   => tenant.test(t.tenant)
        case t: ExactNamespaceMatcher => tenant.test(t.tenant)

case class ExactNamespaceMatcher(tenant: TenantMatcher, namespace: String):
    def test(target: NamespaceMatcher): Boolean = target.matcher match
        case t: AllNamespaceMatcher   => tenant.test(t.tenant)
        case t: ExactNamespaceMatcher => tenant.test(t.tenant) && t.namespace == namespace

case class NamespaceMatcher(matcher: ExactNamespaceMatcher | AllNamespaceMatcher):
    def test(target: NamespaceMatcher): Boolean = matcher match
        case m: ExactNamespaceMatcher => m.test(target)
        case m: AllNamespaceMatcher   => m.test(target)

case class AllTopicMatcher(namespace: NamespaceMatcher):
    def test(target: TopicMatcher): Boolean = target.matcher match
        case t: AllTopicMatcher => namespace.test(t.namespace)
        case t: ExactTopicMatcher => namespace.test(t.namespace)

case class ExactTopicMatcher(namespace: NamespaceMatcher, topic: String):
    def test(target: TopicMatcher): Boolean = target.matcher match
        case t: AllTopicMatcher => namespace.test(t.namespace)
        case t: ExactTopicMatcher => namespace.test(t.namespace) && t.topic == topic

case class TopicMatcher(matcher: AllTopicMatcher | ExactTopicMatcher):
    def test(target: TopicMatcher): Boolean = matcher match
        case m: ExactTopicMatcher  => m.test(target)
        case m: AllTopicMatcher => m.test(target)

case class ResourceMatcher(matcher: TenantMatcher | NamespaceMatcher | TopicMatcher):
    def test(target: ResourceMatcher): Boolean = (matcher, target.matcher) match
        case  (m: TenantMatcher,  t: TenantMatcher) => m.test(t)
        case (m: NamespaceMatcher, t: NamespaceMatcher) => m.test(t)
        case (m: TopicMatcher, t: TopicMatcher) => m.test(t)
